uspace/lib/c/generic/ieee_double.c

```diff
@@ -91,7 +91,12 @@ ieee_double_t extract_ieee_double(double val)
 		/* Denormal or zero. */
 		if (ret.is_denormal) {
 			ret.pos_val.significand = raw_significand;
-			ret.pos_val.exponent = 1 - exponent_bias;
+			if (raw_significand == 0) {
+				ret.pos_val.exponent = -exponent_bias;
+			} else {
+				ret.pos_val.exponent = 1 - exponent_bias;
+			}
+
```

Comment:
This change seems meaningless to me. If significand is 0, the value of exponent doesn't matter.

Reply:
0 is a special case (like infinity and NaN). If exponent != 0 then the hidden bits exist. If exponent == 0, then no hidden bits. Keep in mind that the exponent is underflowed in our case

Reply:
This already is the denormal/zero case. The output is a representation where no bits are hidden.

Reply:
What I mean is, the semantics of `ret` are that the actual numeric value of the number is `pos_val.significand * 2^pos_val.exponent`. If `raw_significand == ret.pos_val.significand == 0`, then `pos_val.exponent` can be arbitrary.

Reply:
but then you have a discrepant between the flag is_denormal and the exponent

Reply:
The flags describes the original IEEE 754 format, it does not affect the semantics of the converted value.

Reply:
I tested it accordingly to the IEEE specification and it's states that the exponent must be 0. I dont think that there is any room for argumentation, either it behaves accordingly to the specification or it doesnt. If you think behaving "similiar" is good enough, fine. But it's wrong.

Reply:
I was under the impression that `ieee_double_t` is purely our own thing. Is the structure and its meaning specified by the IEEE standard?

Reply:
Really? In that case we better call it `helenos_something_nummeric_t`

Reply:
Forgive me if I'm being obtuse, but I truly don't understand what you are talking about. You are saying `0.0` should be represented by `(fp_num_t) { .significand = 0, .exponent = -1075 }`, instead of `(fp_num_t) { .significand = 0, .exponent = -1074 }`. You'll have to explain to me why that matters, because I don't understand.

Reply:
So that you don't misunderstand my intent, please understand that having "standard-looking" interfaces which are not behaving according to standard is something we are very much trying to obliterate from our code base. If you found such a thing, fixing it is very much wanted and appreciated. That being said, I don't understand which part of the standard would be relevant to this `(integer, binary exponent)` tuple representation. As far as I know, no part of the standard even acknowledges existence of such representations, despite their inherent usefulness in math code.

Reply:
> You'll have to explain to me why that matters, because I don't understand.

Because that's not what you would expect to get when you pass `0.0` to that function. The function clearly says it represents the IEEE standard, and that one specifies that  `0.0` => exponent = significant = 0

Or let me rephrase it, why do you think is `-1074` a better choice? Because it's a random number? Perfomance boost?

Reply:
I don't think `-1074` is a better choice. I think adding extra branches just to prefer `-1075` in the place of a number which has no meaning is pointless. IEEE standard specifies the formatting of the _input_ of this function. As I said above, the output is an `(integer, exponent)` tuple, and you are proposing we should specifically prefer `-1075` as the exponent in that tuple.

Reply:
Why would you even expect a specific value there for 0? If it makes difference for your code, the code is just written wrong.

Reply:
In all fairness, I've thought about that too. It collides with other functions though => which is exactly the reason I dont like it to be "random"/undocumented and prefer an extra test to ensure you get an expected result. I can revert that commit, but then you can not meaningful test for the exponent in tests and you must hope that you never have to touch any of the depending functions since you will be left with only one documentation and that will be the source code.

Reply:
If you just want a single concrete value for the tuple, why is `-1075` any better than `-1074`?
It would make sense if you wanted to set it to `0`. Meaningless, but sensible. But that's not what you are proposing.

Reply:
I think you might be confusing the _actual_ exponent with its IEEE _representation_. `double` contains IEEE _representation_ of the exponent, in which all-zero exponent field has a special meaning which could be misinterpreted by applying the bias used for normalized numbers. However, the `fp_pos_t` type used here is the tuple (pair) of `(integer, exponent)`, both of which are their pure mathematical, unbiased value. It has nothing to do with the formats specified by IEEE 754, it's a different format, which is useful for doing math on the value.

Reply:
well I guess that's fair

Reply:
Reverted the commit and corrected tests accordingly (hence remove the test of the exponent in case of `0.0`)